velox/functions/sparksql/DecimalVectorFunctions.cpp

```diff
@@ -0,0 +1,185 @@
+#include "velox/functions/prestosql/ArithmeticImpl.h"
+#include "velox/type/DecimalUtil.h"
+
+namespace facebook::velox::functions::sparksql {
+namespace {
+
+class DecimalRoundFunction : public exec::VectorFunction {
+ public:
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& resultType,
+      exec::EvalCtx& context,
+      VectorPtr& result) const override {
+    auto aType = args[0]->type();
+    auto [aPrecision, aScale] = getDecimalPrecisionScale(*aType);
+    auto [rPrecision, rScale] = getDecimalPrecisionScale(*resultType);
+    int32_t scale = 0;
+    if (args.size() > 1) {
+      VELOX_USER_CHECK(args[1]->isConstantEncoding());
+      scale = args[1]->asUnchecked<ConstantVector<int32_t>>()->valueAt(0);
+    }
+    if (resultType->isShortDecimal()) {
+      if (aType->isShortDecimal()) {
+        applyRoundRows<int64_t, int64_t>(
+            rows,
+            args,
+            aPrecision,
+            aScale,
+            rPrecision,
+            rScale,
+            scale,
+            resultType,
+            context,
+            result);
+      } else {
+        applyRoundRows<int64_t, int128_t>(
+            rows,
+            args,
+            aPrecision,
+            aScale,
+            rPrecision,
+            rScale,
+            scale,
+            resultType,
+            context,
+            result);
+      }
+    } else {
+      if (aType->isShortDecimal()) {
+        applyRoundRows<int128_t, int64_t>(
+            rows,
+            args,
+            aPrecision,
+            aScale,
+            rPrecision,
+            rScale,
+            scale,
+            resultType,
+            context,
+            result);
+      } else {
+        applyRoundRows<int128_t, int128_t>(
+            rows,
+            args,
+            aPrecision,
+            aScale,
+            rPrecision,
+            rScale,
+            scale,
+            resultType,
+            context,
+            result);
+      }
+    }
+  }
+
+  bool supportsFlatNoNullsFastPath() const override {
+    return true;
+  }
+
+ private:
+  template <typename R /* Result type */>
+  R* prepareResults(
+      const SelectivityVector& rows,
+      const TypePtr& resultType,
+      exec::EvalCtx& context,
+      VectorPtr& result) const {
+    context.ensureWritable(rows, resultType, result);
+    result->clearNulls(rows);
+    return result->asUnchecked<FlatVector<R>>()->mutableRawValues();
+  }
+
+  template <typename R /* Result */, typename A /* Argument */>
+  void applyRoundRows(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      uint8_t aPrecision,
+      uint8_t aScale,
+      uint8_t rPrecision,
+      uint8_t rScale,
+      int32_t scale,
+      const TypePtr& resultType,
+      exec::EvalCtx& context,
+      VectorPtr& result) const {
+    // Single-arg deterministic functions receive their only
+    // argument as flat or constant only.
```

Comment:
add a check to ensure this?

Reply:
Yes, we will check the argument number here https://github.com/facebookincubator/velox/pull/5597/files#diff-eac35f6f71f4d48a5603fa4ebe182c42d18202eea76f8b5a105ce6612163afe0R89

Reply:
This check gets from https://github.com/oap-project/velox/blob/main/velox/functions/prestosql/DecimalVectorFunctions.cpp#L128, I don't know where to ensure this.

Reply:
Added the check.